intel_adsp: power: Fix for ACE30 Power Management

tmleman opened the pull request:
# Summary of Changes for Pull Request

This pull request includes a series of updates to the ACE30 PTL firmware and device tree source files. The changes improve the accuracy of the hardware representation, clean up redundant code, ensure proper power management procedures are followed, and fix issues found in power flows. Below is a summary of the included changes:

## Device Tree Source (DTS) Updates

### Reordering of Power Domains
- Power domain entries in the ACE30 PTL DTS file have been reordered by bit position for improved readability and logical grouping.

### Removal of Non-Existent Power Domains
- Removed definitions of power domains from the ACE30 PTL DTS file that do not exist in the actual hardware, preventing potential misconfiguration.

### Removal of ALH Nodes
- Removed the Audio Link Hub (ALH) nodes from the ACE30 PTL DTS file as they are not utilized in the ACE 3.0 PTL architecture.

## Firmware Code Updates

### Use of MMU Reinitialization API
- Updated the core context restore process to use the new `xtensa_mmu_reinit()` API instead of `xtensa_mmu_init()`. This change fixes an issue in the power flow where the MMU context was not properly preserved across low-power states, ensuring that runtime changes to the MMU configuration are retained.

iganakov committed:
intel_adsp: ace30: Correct power control register bitfield definitions

This patch updates the power control and status register bitfield
definitions in the ACE30 PTL ADSP power management header to match the
documented hardware specifications. The previous definitions contained
discrepancies that did not align with the actual hardware layout,
potentially leading to incorrect assumptions and usage within the
firmware.

Changes made in this patch:
- Renamed 'rsvd0' to 'rsvd4' to accurately represent the reserved bits
  starting at bit position 4.
- Removed the 'rsvd6' field, which was incorrectly defined and is not
  present in the hardware register layout.
- Adjusted the bit widths for 'ioxpgs' and 'mlpgs' to correctly reflect
  the number of bits these fields occupy in the hardware.
- Introduced a new 'rsvd15' field in both 'ace_pwrctl2' and
  'ace_pwrsts2' structures to account for the remaining reserved bits,
  ensuring the structure sizes accurately represent the full register
  width.

By correcting these bitfield definitions, the firmware's power
management code will now be consistent with the actual hardware design,
improving reliability and maintainability of the codebase.

Signed-off-by: Ievgen Ganakov <ievgen.ganakov@intel.com>
Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman committed:
intel_adsp/ace: power: Use MMU reinit API on core context restore

Replace the MMU initialization call with the new MMU re-initialization
API during the core context restore process in the ACE power management
code.

The previous code was directly calling `xtensa_mmu_init()` upon
restoring the core context, which is not appropriate when the MMU
context may have been preserved during low-power states. The new
`xtensa_mmu_reinit()` API is designed to re-establish the MMU context
without overwriting the existing page table, ensuring that any runtime
changes to the MMU configuration are retained.

Changes made in this patch:
- Removed the call to `xtensa_mmu_init()` from the
  `_restore_core_context()` function.
- Added a call to `xtensa_mmu_reinit()` after restoring the
  miscellaneous registers.

This update aligns the ACE power management code with the correct MMU
handling procedures when recovering from low-power states, as per the
recent changes in the Xtensa MMU support.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

andyross commented on dts/xtensa/intel/intel_adsp_ace30_ptl.dtsi:
```diff
@@ -439,22 +439,10 @@
 			};
-			ml1_domain: ml1_domain {
-				compatible = "intel,adsp-power-domain";
-				bit-position = <13>;
```
Nitpick: you just went through the trouble to adjust the bit-position field three patches back, and now you're removing it?  Doing it in the opposite order would be clearer, with less churn.

tmleman replied:
This is a valid point :) For some reason, I planned to organize these power domains first and clean them up later. In my defense, it's already Friday 5:46 PM in my time zone.

Best regards.

andyross reviewed (approved):
Seems like good cleanup.  One nit.

tmleman committed:
dts: xtensa: intel: Reorder power domains by bit position in ACE30

Rearrange the power domain entries in the ACE30 PTL device tree source
file to be in ascending order according to their bit positions. This
reordering improves the readability of the device tree source by
grouping power domains logically according to their bit position within
the power management registers.

The changes in this patch include:
- Moving 'ml1_domain' and 'ml0_domain' to their correct positions
  according to their bit-position values (13 and 12, respectively).
- Adjusting the order of 'io3_domain', 'io2_domain', 'io1_domain', and
  'io0_domain' to reflect their bit positions (11, 10, 9, and 8).
- Placing 'hub_hp_domain' and 'hst_domain' at their new positions
  according to their bit-position values (6 and 5).

No functional changes are introduced with this patch. It solely aims to
make the device tree source more intuitive and easier to navigate when
mapping power domains to their respective control bits.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman committed:
dts: xtensa: intel: Remove ALH nodes from ACE 3.0 PTL DTS

Remove the Audio Link Hub (ALH) nodes from the ACE 3.0 PTL DTS file.

This patch cleans up the Device Tree Source (DTS) for the ACE 3.0 PTL
platform by removing the definitions of the ALH DAI nodes. The ALH
interface is not utilized in the ACE 3.0 PTL architecture, making these
nodes redundant.

The following changes are made:
- Deleted the 'alh0' and 'alh1' nodes, which were previously defined
  with FIXME comments indicating a problematic modeling of individual
  ALH channels/instances using node labels.

This cleanup helps to prevent confusion and potential errors in device
configuration by ensuring that the DTS reflects the actual hardware
capabilities of the ACE 3.0 PTL platform.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman committed:
dts: xtensa: intel: Remove non-existent power domains from ACE30 PTL DTS

This patch removes definitions of power domains from the ACE30 PTL DTS
file that do not exist in the actual hardware.

The following power domain nodes have been removed:
- 'ml1_domain' with a bit-position of <13>
- 'io3_domain' with a bit-position of <11>
- 'io2_domain' with a bit-position of <10>

These nodes were previously included in the DTS file but do not
correspond to any physical power domain in the ACE30 PTL hardware. Their
presence in the DTS could lead to confusion and misconfiguration, as the
software might attempt to interact with non-existent hardware features.

By removing these nodes, the DTS now accurately reflects the hardware
capabilities of the ACE30 PTL platform, ensuring that the power
management infrastructure within the firmware operates based on the
correct hardware configuration.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>
